.github/workflows/docker-dev.yaml

```diff
@@ -0,0 +1,66 @@
+name: Build and push development images
+
+on:
+  push:
+    branches:
+      - 'main'
+      - 'next'
+
+jobs:
+  docker:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Cache Docker layers
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
+      - name: Login to DockerHub
+        uses: docker/login-action@v1 
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_TOKEN }}
+      - name: Extract branch name
+        id: extract_branch
+        run: echo "##[set-output name=branch;]$(echo ${GITHUB_REF#refs/heads/})"
+      - name: Build binary
+        env:
+          TAG: ${{ steps.extract_branch.outputs.branch }}
+          COMMIT: ${{ github.sha }}
+          REPO_INFO: https://github.com/${{ github.repository }}.git
+        id: docker_build_binary
+        uses: docker/build-push-action@v2
+        with:
+          push: false
+          file: Dockerfile
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache
+          target: build
+      - name: Build and push Alpine
+        id: docker_build_alpine
+        env:
+          TAG: ${{ steps.extract_branch.outputs.branch }}
+        uses: docker/build-push-action@v2
+        with:
+          push: true
+          file: Dockerfile
+          tags: |
+            kong/kubernetes-ingress-controller:${{ steps.extract_branch.outputs.branch }}-alpine
+            kong/kubernetes-ingress-controller:${{ steps.extract_branch.outputs.branch }}
+          cache-from: type=local,src=/tmp/.buildx-cache
+          target: alpine
+      - name: Build and push RedHat
+        id: docker_build_redhat
+        env:
+          TAG: ${{ steps.extract_branch.outputs.branch }}
+        uses: docker/build-push-action@v2
+        with:
+          push: true
+          file: Dockerfile
+          tags: kong/kubernetes-ingress-controller:${{ steps.extract_branch.outputs.branch }}-rhel
```

Comment:
One last question.

We seem to be in agreement that development images should not bloat the user-facing image repository with tags and that these should go to a separate repo.

At the same time, `main` and `next` are clearly development tags - it just happens that more will follow as we improve our CI.

Maybe the right moment to establish that separate repo for dev tags is now (so that `main` and `next` go there)? If, otherwise, we do this later, we'll need to remove `main` and `next` tags from this (user-facing) repo.

Reply:
See https://github.com/Kong/kubernetes-ingress-controller/pull/1156#discussion_r610919251

They are a limited set of pre-release tags for branches that contain code that will eventually be released. They're there if users wish to test new features and/or stage configuration in their dev environments that requires unreleased functionality so that they can more quickly promote to prod once a release tag is cut. There are only two of them and it's easy to understand what they represent, whereas the individual ref builds used in CI are many, require you know which hash corresponds to a particular feature, and may not be fully reviewed.

Reply:
The way I see the state we're heading towards is:

1. some docker registry A has release tags `1.0`, `1.1.1`, `2.0`, `2.0.0-alpha.1`, `2.0.0-beta.3` etc that exist there indefinitely
2. some docker registry B has tags created from git SHAs, from every build - that are immutable, but are pruned after some time (e.g. a month)
3. some docker registry C has tags created from branch names `main`, `next`, `feat/awesome-feature`, etc etc etc which are a (moving) alias for the git SHAs above

I was thinking that `kong/kubernetes-ingress-controiller` makes sense for A, and something like `kong/kubernetes-ingress-controller-dev` for B and C.

Would you like to (in the long run) replicate `main` and `next` into A as well? Or split C into `main`/`next` going into one repo, and all branches BUT these two going to another repo?

Reply:
> Would you like to (in the long run) replicate main and next into the "user-facing" repo as well?

This is what I intended to do from the outset.

Ultimately it's not difficult to add an additional push target--you just add another item under the `tags` section that uses another repo, so I wouldn't worry about this for now.